src/controller/java/src/matter/controller/MatterControllerImpl.kt

```diff
@@ -0,0 +1,661 @@
+                          timeStamp = timestamp,
+                          data = event.tlvValue
+                        )
+                      successes.add(readData)
+                    }
+                  }
+                }
+              }
+
+              trySendBlocking(SubscriptionState.NodeStateUpdate(ReadResponse(successes, failures)))
+                .onFailure { ex ->
+                  logger.log(Level.SEVERE, "Error sending NodeStateUpdate to subscriber: %s", ex)
+                }
+            }
+
+            override fun onError(
+              attributePath: ChipAttributePath?,
+              eventPath: ChipEventPath?,
+              ex: Exception
+            ) {
+              attributePath?.let {
+                logger.log(Level.INFO, "Report error for attributePath:%s", it.toString())
+                val tmpAttributePath: AttributePath = attributePath.wrap()
+                val attributeFailure = ReadFailure.Attribute(path = tmpAttributePath, error = ex)
+                failures.add(attributeFailure)
+              }
+              eventPath?.let {
+                logger.log(Level.INFO, "Report error for eventPath:%s", it.toString())
+                val tmpEventPath: EventPath = eventPath.wrap()
+                val eventFailure = ReadFailure.Event(path = tmpEventPath, error = ex)
+                failures.add(eventFailure)
+              }
+
+              // The underlying subscription is terminated if both attributePath & eventPath are
+              // null
+              if (attributePath == null && eventPath == null) {
+                trySendBlocking(
+                    SubscriptionState.SubscriptionErrorNotification(CHIP_ERROR_UNEXPECTED_EVENT)
+                  )
+                  .onFailure { exception ->
+                    logger.log(
+                      Level.SEVERE,
+                      "Error sending SubscriptionErrorNotification to subscriber: %s",
+                      exception
+                    )
+                  }
+              }
+            }
+
+            override fun onDone() {
+              logger.log(Level.INFO, "Subscription update completed")
+            }
+          }
+
+        deviceController.subscribeToPath(
+          subscriptionEstablishedHandler,
+          resubscriptionAttemptHandler,
+          reportHandler,
+          devicePtr,
+          attributePaths,
+          eventPaths,
+          request.minInterval.seconds.toInt(),
+          request.maxInterval.seconds.toInt(),
+          request.keepSubscriptions,
+          request.fabricFiltered,
+          CHIP_IM_TIMEOUT_MS
+        )
+
+        awaitClose { logger.log(Level.FINE, "Closing flow") }
+      }
+      .buffer(capacity = UNLIMITED)
+  }
+
+  override suspend fun read(request: ReadRequest): ReadResponse {
+    // To prevent potential issues related to concurrent modification, assign
+    // the value of the mutable property 'nodeId' to a temporary variable.
+    val nodeId = this.nodeId
```

Comment:
here this seems not needed: we use this right away in the first call. Even if suspend, there is no way this can race right?

Reply:
It is needed, otherwise, we will get error "'nodeId' is a mutable property that could have been changed by this time"